keps/sig-scheduling/5142-pop-backoffq-when-activeq-empty/README.md

```diff
@@ -0,0 +1,409 @@
+
+In case of API calls errors (e.g. network issues), backoffQ allows to limit number of retries in a short term.
+This proposal will take those pods earlier, leading to losing this mechanism.
+
+After merging [kubernetes#128748](github.com/kubernetes/kubernetes/pull/128748), 
+it will be possible to distinguish pods backing off because of errors from those backing off because of unschedulable attempt.
+This information could be used when popping, by filtering only the pods that are from unschedulable attempt or even splitting backoffQ.
+
+This has to be resolved before the beta is released.
+
+#### One pod in backoffQ could starve the others
+
+The head of BackoffQ is the pod with the closest backoff expiration,
+and the backoff time is calculated based on the number of scheduling failures that the pod has experienced.
+If one pod has a smaller attempt counter than others,
+could the scheduler keep popping this pod ahead of other pods because the pod's backoff expires faster than others?
+Actually, that wouldn't happen because the scheduler would increment the attempt counter of pods from backoffQ as well,
+which would make the backoff time of pods bigger every after the scheduling attempt,
+and the pod that had a smaller attempt number eventually won't be popped out.
+
+## Design Details
+
+### Popping from backoffQ in activeQ's pop()
+
+To achieve the goal, activeQ's `pop()` method needs to be changed:
+1. If activeQ is empty, then instead of waiting for a pod to arrive at activeQ, popping from backoffQ is tried.
+2. If backoffQ is empty, then `pop()` is waiting for pod as previously.
+3. If backoffQ is not empty, then the pod is processed like the pod would be taken from activeQ, including increasing attempts number.
+   It is poping from a heap data structure, so it should be fast enough not to cause any performance troubles.
+
+To support monitoring, when popping from backoffQ, 
+the `scheduler_queue_incoming_pods_total` metric with an `activeQ` queue and a new `PopFromBackoffQ` event label will be incremented.
+
+### Notifying activeQ condition when new pod appears in backoffQ
+
+Pods might appear in backoffQ while `pop()` is hanging on point 2. 
+That's why it will be required to call `broadcast()` on condition after adding a pod to backoffQ.
+It shouldn't cause any performance issues.
+
+We could eventually want to move backoffQ under activeQ's lock, but it's out of scope of this KEP.
+
+### Calling PreEnqueue for backoffQ
+
+`PreEnqueue` plugins have to be called for every pod before they are taken to scheduling cycle.
+Initially, those plugins were called before moving pod to activeQ.
+With this proposal, `PreEnqueue` will need to be called before moving pod to backoffQ 
+and those calls need to be skipped for the pods that are moved later from backoffQ to activeQ.
+At moveToActiveQ level, these two paths could be distinguished by checking if event is equal to `BackoffComplete`.
```

Comment:
let's mention the motivation of this change

```suggestion
Currently, we call `PreEnqueue` at every time pods are being moved to activeQ. 
But, with this proposal, `PreEnqueue` will be called before moving pod to backoffQ, not when popping pods directly from backoffQ.
Otherwise, a direct popping would be inefficient; it has to take the top backoffQ pod, check if it goes through PreEnqueue plugins, if not check the next backoffQ pod... until it finds the pod that goes through all PreEnqueue plugins.

It means we'd have two paths that PreEnqueue plugins are invoked; when new pods are created and entering to the scheduling queue and when pods are pushed in backoffQ.
At moveToActiveQ level, these two paths could be distinguished by checking if event is equal to `BackoffComplete`.
```